workers/common/src/main/kotlin/common/env/EnvironmentService.kt

```diff
@@ -177,9 +218,72 @@ class EnvironmentService(
      */
     private fun assignServicesToOrtRun(context: WorkerContext, services: Collection<InfrastructureService>) {
         services.forEach { service ->
-            infrastructureServiceRepository.getOrCreateForRun(service, context.ortRun.id)
+            dynamicInfrastructureServiceRepository.getOrCreateForRun(
+                service.mapToDynamicInfrastructureService(),
+                context.ortRun.id
+            )
         }
     }
+
+    /**
+     * Resolve a secret by its [secretName] for the current [ortRun]. The secret is searched in the repository, product,
+     * and organization levels, in that order. If a secret is found at multiple levels, the one at the repository level
+     * is preferred, followed by the product level, and finally the organization level. In case a secret is found at
+     * multiple levels, a warning is logged about this ambiguity. If no secret is found at all, return null.
+     */
+    fun resolveSecretByName(secretName: String, ortRun: OrtRun, serviceName: String): Secret? {
+        // Try to get from all 3 levels: repository, product, and organization
+        val secretRepositoryLevel =
+            secretRepository.getByIdAndName(RepositoryId(ortRun.repositoryId), secretName)
+        val secretProductLevel =
+            secretRepository.getByIdAndName(ProductId(ortRun.productId), secretName)
+        val secretOrganizationLevel =
+            secretRepository.getByIdAndName(OrganizationId(ortRun.organizationId), secretName)
+
+        secretRepositoryLevel?.let {
+            secretProductLevel?.let {
+                secretOrganizationLevel?.let {
+                    logger.warn(
```

Comment:
It is certainly helpful to log from which level the secret was used, but I would not use warning level. Overriding a secret may be a normal use case.